### DISCUSSION TITLE
Async scoped session vs. a "scoped" engine

### DISCUSSION THREAD

**Discussion by pansen** [sqlalchemy/sqlalchemy#7588] [Usage Questions] [answered]
Dear all, 

first I want to mention, how much SQLAlchemy is a wonderful project. I really enjoy using that, everytime I use it. 

After switching to the async API, one thing is still unclear for me. In the [documentation](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#) it is mentioned, that an engine has to be bound to an `asyncio` loop (otherwise there is this `got Future attached to a different loop` error). 

A [scoped async session](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-asyncio-scoped-session) is scoped per `asyncio` loop, as far as I understand - yet it shares the same engine. 

Is there something like *scoped engine*, which takes care about the tracking of "is there an engine available for the current loop, which one is currently in"?

Here is a minimal example

```python
from asyncio import current_task

from sqlalchemy.ext.asyncio import AsyncSession, async_scoped_session
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.asyncio import AsyncEngine, create_async_engine

async_session_factory = sessionmaker(class_=AsyncSession)
AsyncSessionMaker = async_scoped_session(async_session_factory, scopefunc=current_task)


async def run(self):
    _engine = create_async_engine(
        config.SQLA_URL,
        echo=False,
        # TODO: fix this to not use `NullPool`
        poolclass=NullPool,
    )
    AsyncSessionMaker.configure(bind=_engine)

    while True:
        async with AsyncSessionMaker(expire_on_commit=False) as session:
            # do stuff
            await session.commit()

    await _engine.dispose()
    

async def handler(self):
    """
    This handler might be called from a different loop. 
    """
    async with AsyncSessionMaker(expire_on_commit=False) as session:
        # do stuff
        await session.commit()
```

The question is: what to do in `handler` to ensure an engine per that loop, which is calling. 

Cheers!

**Comment by zzzeek** [comment, marked as answer]
> Dear all,
> 
> first I want to mention, how much SQLAlchemy is a wonderful project. I really enjoy using that, everytime I use it.
> 
> After switching to the async API, one thing is still unclear for me. In the [documentation](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#) it is mentioned, that an engine has to be bound to an `asyncio` loop (otherwise there is this `got Future attached to a different loop` error).

i think this only happens if you try to use the engine before the event loop is started.  we should have checks in place that keep it from prematurely associating itself with the wrong event loop at this point, so if your application is using normal asyncio, there shouldn't be any problem.  the docs I see refer to some approach that would have multiple event loops in place somehow (@CaselIT is there a concrete example we can link towards when we say "combining asyncio with multithreading" ?  is that a thing people do?  not really sure why one would do that given the GIL...)

> 
> A [scoped async session](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-asyncio-scoped-session) is scoped per `asyncio` loop, as far as I understand - yet it shares the same engine.

are you actually using multiple asyncio loops simultaneously?   I'm not familiar with why one would be doing that.

> 
> Is there something like _scoped engine_, which takes care about the tracking of "is there an engine available for the current loop, which one is currently in"?

I'm not a big fan of the "scoped registry" thing, even though it was my idea many years ago, because I find that it confuses people a lot.   it falls under a category of patterns of "useful pattern, provided the team using it has implemented it themselves so they know exactly what they are doing".    for example, if someone is writing a web app in a threaded system, but they dont have any understanding of what a threading.local() is, I dont want that person using ScopedSession because their confusion over how threading.local() works will instead become bug reports on my end :) . id rather they understand the concepts.

If you have some infrastructure that is spinning up multiple event loops (need to see that), you can have that same system also be responsible for creating async engines per loop and passing them in to the top level awaitable.  if you want the awaitables to have a global registry of sorts, then yes, you'd make your own "scoped registry", which in the old days would be just a dictionary keyed to something like get_event_loop() and the values are async engine objects, however today you should probably use Python stdlib [contextvars](https://docs.python.org/3/library/contextvars.html) which support asyncio and for which our own greenlet system has explicit support, so that system works correctly.

**Reply by pansen** [reply]
> i think this only happens if you try to use the engine before the event loop is started. we should have checks in place that keep it from prematurely associating itself with the wrong event loop at this point, so if your application is using normal asyncio, there shouldn't be any problem. 

Yes, I did not make that clear, but I do start my part of the application in a thread, since a library, which I'm using occupies the main `asyncio` loop. In that thread, I start with what you mention

```python
    def _run_in_thread(thing: AbstractThing):
        # As we are called in a thread, we need to create an eventloop
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)

        log.info('Starting %s in thread ...', bc)
        asyncio.run(thing.run())
```

Further down the line, that library is calling handlers of `AbstractThing`, which leads to two loops interacting somehow. 

> I'm not a big fan of the "scoped registry" thing, even though it was my idea many years ago, because I find that it confuses people a lot. 

Definitely, I'm affected as well :D - I always need to rethink that, if there is a new project, where this needs to be setup (even though most of the time the scoped session is only used for tests in my case).

> If you have some infrastructure that is spinning up multiple event loops (need to see that), you can have that same system also be responsible for creating async engines per loop and passing them in to the top level awaitable. if you want the awaitables to have a global registry of sorts, then yes, you'd make your own "scoped registry", which in the old days would be just a dictionary keyed to something like get_event_loop() and the values are async engine objects, however today you should probably use Python stdlib contextvars which support asyncio and for which our own greenlet system has explicit support, so that system works correctly.

Thanks for explaining that, this is what I'm going to do now :) I first wanted to be sure if I miss something before there will be a custom solution. 

✌🏼

**Reply by pansen** [reply]
FTR maybe I'm also rethinking that whole thread thing and try to integrate into the main loop. After reading your message again.

**Reply by CaselIT** [reply]
> (@CaselIT is there a concrete example we can link towards when we say "combining asyncio with multithreading" ? is that a thing people do? not really sure why one would do that given the GIL...)

No clue specifically, I guess it can happen, but probably we can place a note in the docs mentioning that's not an usual case?

**Reply by zzzeek** [reply]
i would just say, "in the uncommon case that..."
